### ISSUE TITLE
Add support for fast caching on github self-hosted runners

### ISSUE THREAD

**Issue by gunta** [google/wireit#239]
## Explainer
As we know, Wireit is likely going to remove dependency on github cache action, so that would be a good timing to handle this issue.
Wireit introduces heavy caching which is perfect when you want to save CPU cycles, but its trade-off is network speed and hard-disk read/write speed.

One of the biggest reasons to use a self-hosted runner is that you can provide a build machine with 10x more core/memory to make builds faster. 

### Run actions/cache@v3 download cache
| Actions Runner | vCPU Spec | RAM Spec | Datacenter | Cache Download Speed |
| --------------- | ----------- | ----------- | ----------- | ------------------------ |
| [Official Github.com Runner](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources) | 2 | 7 GB | Azure | 155 MBs/sec
| [BuildJet Runner](https://buildjet.com) | 16 | 64 GB | Their own datacenter | 20 MBs/sec
| Self-hosted Runner | 16 | 160 GB | Private datacenter | 7 MBs/sec

Since cached files are big, this makes files **transfer speed time being > slower** than the CPU cycles saved by wireit.

In short: the cache and wireit have to be `on the same network/datacenter` in order to provide a total performance benefit.

## Requirements
- [ ] Add support for GitHub self-hosted runners caching
  - [ ] S3 API (most self-hosted runners can provide a compatible S3 API for storing and retrieving cached files) `@aws-sdk/client-s3`
  - [ ] Cloud Storage API `@google-cloud/storage`
  - [ ] Google Cloud Build: https://github.com/google/wireit/issues/145
  - [ ] Any other popular way to store cache next to the database: Something like https://github.com/ItalyPaleAle/SMCloudStore https://github.com/minio/minio

**Comment by gunta** [comment]
Related: https://github.com/google/wireit/issues/107

**Comment by aomarks** [comment]
Hello, thanks for the issue!

> As we know, [Wireit as removed the dependency on github cache action](https://github.com/google/wireit/commit/0e93dcfed9ca9c613c2e230d829b8aceda640cc6), and this introduces performance regressions.

Interesting -- this PR should have been a drop-in replacement for the client library. Wireit still supports exactly the same integration with GitHub Actions Caching. Did you experience a performance regression after upgrading?

> ### Run action download cache

Are the download speeds in this table correct? Is it possible the "Cache Download Speed" column is reversed?

> Since cached files are big, this makes files **transfer speed time being > slower** than the CPU cycles saved by wireit.
> 
> In short: the cache and wireit have to be `on the same network/datacenter` in order to provide a total performance benefit.

I haven't found this to be true in my own experience so far -- but it's possible you have a different bytes to runtime ratio. In the places we have adopted Wireit, it seems significantly faster to use GitHub Actions Caching than to run the build script itself. Some scripts are more CPU intensive than others so it might *always* be a win, but it generally seems to be.

Could you provide any more details about the kinds of scripts where you are seeing better performance by re-generating versus pulling from cache?

> [ ] Add support for GitHub self-hosted runners caching
> * [ ]  S3 API (most self-hosted runners can provide a compatible S3 API for storing and retrieving cached files) `@aws-sdk/client-s3`
>         
>       
>   
>       Create an issue with the title S3 API (most self-hosted runners can provide a compatible S3 API for storing and retrieving cached files) <code class="notranslate">@aws-sdk/client-s3`. Press Enter to convert to an issue instantly. Press Alt-Enter to open the create new issue form in the current tab. Press Shift-Enter to open the create new issue form.
> * [ ]  Cloud Storage API `@google-cloud/storage`
>         
>       
>   
>       Create an issue with the title Cloud Storage API <code class="notranslate">@google-cloud/storage`. Press Enter to convert to an issue instantly. Press Alt-Enter to open the create new issue form in the current tab. Press Shift-Enter to open the create new issue form.
> * [ ]  Google Cloud Build: [Caching for Google Cloud Build #145](https://github.com/google/wireit/issues/145)
> * [ ]  Any other popular way to store cache next to the database: Something like https://github.com/ItalyPaleAle/SMCloudStore https://github.com/minio/minio

Definitely open to supporting other caching backends. I think we should do this via some sort of plugin system.

**Comment by gunta** [comment]
> Interesting -- this PR should have been a drop-in replacement for the client library that communicates with the GitHub Actions Caching service. Wireit still supports exactly the same integration with GitHub Actions Caching. Did you experience a performance regression after upgrading?

The problem is that we were using a forked `github/actions` in order to support S3 API transfer to our hosted runners.

> Are the download speeds in this table correct? Is it possible the "Cache Download Speed" column is reversed?
Sadly, it is correct. 

Here is some example from the build logs

# Github.com Runner (Azure @ US Region)
```
Run actions/cache@v3
Received 281018368 of 354678885 (79.2%), 267.7 MBs/sec
Received 354678885 of 354678885 (100.0%), 231.2 MBs/sec
Cache Size: ~338 MB (354678885 B)
```
**Total: 33s**

# BuildJet Runner (Owned Datacenter likely @ US Region)
```
Run actions/cache@v3
Received 0 of 374877461 (0.0%), 0.0 MBs/sec
Received 4194304 of 374877461 (1.1%), 2.0 MBs/sec
Received 16777216 of 374877461 (4.5%), 5.3 MBs/sec
Received 41943040 of 374877461 (11.2%), 10.0 MBs/sec
Received 62914560 of 374877461 (16.8%), 12.0 MBs/sec
Received 88080384 of 374877461 (23.5%), 14.0 MBs/sec
Received 109051904 of 374877461 (29.1%), 14.9 MBs/sec
Received 134217728 of 374877461 (35.8%), 16.0 MBs/sec
Received 150994944 of 374877461 (40.3%), 16.0 MBs/sec
Received 171966464 of 374877461 (45.9%), 16.4 MBs/sec
Received 197132288 of 374877461 (52.6%), 17.1 MBs/sec
Received 218103808 of 374877461 (58.2%), 17.3 MBs/sec
Received 247463936 of 374877461 (66.0%), 18.1 MBs/sec
Received 272629760 of 374877461 (72.7%), 18.6 MBs/sec
Received 289406976 of 374877461 (77.2%), 18.4 MBs/sec
Received 314572800 of 374877461 (83.9%), 18.7 MBs/sec
Received 335544320 of 374877461 (89.5%), 18.8 MBs/sec
Received 364904448 of 374877461 (97.3%), 19.3 MBs/sec
Received 374877461 of 374877461 (100.0%), 19.0 MBs/sec
Cache Size: ~358 MB (374877461 B)
```
**Total: 1m 5s**

# Self-hosted Runner (Private Datacenter @ Tokyo Region)
```
Run actions/cache@v3
Received 0 of 375180421 (0.0%), 0.0 MBs/sec
Received 0 of 375180421 (0.0%), 0.0 MBs/sec
Received 0 of 375180421 (0.0%), 0.0 MBs/sec
Received 0 of 375180421 (0.0%), 0.0 MBs/sec
Received 12582912 of 375180421 (3.4%), 2.4 MBs/sec
Received 25165824 of 375180421 (6.7%), 4.0 MBs/sec
Received 25165824 of 375180421 (6.7%), 3.4 MBs/sec
Received 25165824 of 375180421 (6.7%), 3.0 MBs/sec
Received 29360128 of 375180421 (7.8%), 3.1 MBs/sec
Received 37748736 of 375180421 (10.1%), 3.6 MBs/sec
Received 50331648 of 375180421 (13.4%), 4.4 MBs/sec
Received 58720256 of 375180421 (15.7%), 4.7 MBs/sec
Received 58720256 of 375180421 (15.7%), 4.3 MBs/sec
Received 58720256 of 375180421 (15.7%), 4.0 MBs/sec
Received 67108864 of 375180421 (17.9%), 4.3 MBs/sec
Received 67108864 of 375180421 (17.9%), 4.0 MBs/sec
Received 71303168 of 375180421 (19.0%), 4.0 MBs/sec
Received 83886080 of 375180421 (22.4%), 4.4 MBs/sec
Received 83886080 of 375180421 (22.4%), 4.2 MBs/sec
Received 83886080 of 375180421 (22.4%), 4.0 MBs/sec
Received 92274688 of 375180421 (24.6%), 4.2 MBs/sec
Received 100663296 of 375180421 (26.8%), 4.4 MBs/sec
Received 104857600 of 375180421 (27.9%), 4.3 MBs/sec
Received 109051904 of 375180421 (29.1%), 4.3 MBs/sec
Received 113246208 of 375180421 (30.2%), 4.3 MBs/sec
Received 125829120 of 375180421 (33.5%), 4.6 MBs/sec
Received 130023424 of 375180421 (34.7%), 4.6 MBs/sec
Received 130023424 of 375180421 (34.7%), 4.4 MBs/sec
Received 134217728 of 375180421 (35.8%), 4.4 MBs/sec
Received 150994944 of 375180421 (40.2%), 4.8 MBs/sec
Received 155189248 of 375180421 (41.4%), 4.8 MBs/sec
Received 159383552 of 375180421 (42.5%), 4.7 MBs/sec
Received 159383552 of 375180421 (42.5%), 4.6 MBs/sec
Received 167772160 of 375180421 (44.7%), 4.7 MBs/sec
Received 184549376 of 375180421 (49.2%), 5.0 MBs/sec
Received 184549376 of 375180421 (49.2%), 4.9 MBs/sec
Received 188743680 of 375180421 (50.3%), 4.9 MBs/sec
Received 188743680 of 375180421 (50.3%), 4.7 MBs/sec
Received 205520896 of 375180421 (54.8%), 5.0 MBs/sec
Received 209715200 of 375180421 (55.9%), 5.0 MBs/sec
Received 209715200 of 375180421 (55.9%), 4.9 MBs/sec
Received 218103808 of 375180421 (58.1%), 5.0 MBs/sec
Received 226492416 of 375180421 (60.4%), 5.0 MBs/sec
Received 234881024 of 375180421 (62.6%), 5.1 MBs/sec
Received 234881024 of 375180421 (62.6%), 5.0 MBs/sec
Received 234881024 of 375180421 (62.6%), 4.9 MBs/sec
Received 243269632 of 375180421 (64.8%), 4.9 MBs/sec
Received 260046848 of 375180421 (69.3%), 5.2 MBs/sec
Received 268435456 of 375180421 (71.5%), 5.2 MBs/sec
Received 268435456 of 375180421 (71.5%), 5.1 MBs/sec
Received 268435456 of 375180421 (71.5%), 5.0 MBs/sec
Received 272629760 of 375180421 (72.7%), 5.0 MBs/sec
Received 276824064 of 375180421 (73.8%), 5.0 MBs/sec
Received 285212672 of 375180421 (76.0%), 5.0 MBs/sec
Received 293601280 of 375180421 (78.3%), 5.1 MBs/sec
Received 293601280 of 375180421 (78.3%), 5.0 MBs/sec
Received 306184192 of 375180421 (81.6%), 5.1 MBs/sec
Received 310378496 of 375180421 (82.7%), 5.1 MBs/sec
Received 318767104 of 375180421 (85.0%), 5.2 MBs/sec
Received 322961408 of 375180421 (86.1%), 5.1 MBs/sec
Received 335544320 of 375180421 (89.4%), 5.2 MBs/sec
Received 335544320 of 375180421 (89.4%), 5.2 MBs/sec
Received 343932928 of 375180421 (91.7%), 5.2 MBs/sec
Received 343932928 of 375180421 (91.7%), 5.1 MBs/sec
Received 352321536 of 375180421 (93.9%), 5.2 MBs/sec
Received 366791813 of 375180421 (97.8%), 5.3 MBs/sec
Received 375180421 of 375180421 (100.0%), 5.4 MBs/sec
Cache Size: ~358 MB (375180421 B)
```
**Total: 3m 3s** 

> Some scripts are more CPU intensive than others so it might not always be a win, but it generally seems to be.

So here is why: Wireit is having a very good impact in cache performance overall, however we're using CPU-bounded but heavily optimized stuff written in Rust like [Dprint prettier](https://github.com/dprint/dprint-plugin-prettier) (which is orders of magnitudes faster than the official Prettier) so the total running time with a clean cache even happens to be just some seconds. I think this is gonna be more the case once most tools move to natively optimized systems (like [bun](https://bun.sh), Deno, etc).

But restoring the dependencies and build cache of 358 MB in our datacenters is over 3 min (vs. 33seconds of Github.com Runner), so now the bottleneck stopped to be CPU and happens to be Network I/O.

By the way, this was the reply from BuildJet

> Unfortunately, workflows that heavily rely on the download GitHub caches, will be slower than GitHub Action's native runner, simply due to our distance to the GA storage machines compared to GitHub's own runners. 
With that being said, we're looking into fixing this by providing a BuildJet Specific caching solution that, which will be much faster than GitHub’s own caching solution.
This will unfortunately take some time and I can't give you a precise release date. 

> Definitely open to supporting other caching backends. I think we should do this via some sort of plugin system.

Looking forward to it.

**Comment by aomarks** [comment]
Please re-open if you have additional information, but from the description so far I'm not sure that it's clear that there was actually a regression here. The new GitHub Caching implementation should have been a drop-in replacement for the library we were using before.
